### PR TITLE
Get-DbaLogin, Set-DbaLogin, Connect-SqlInstance - block use with Azure via AzureUnsupported param

### LINKED ISSUE #7004: Set-DBALogin error on Azure SQL Database
#### Host used

- [X] powershell.exe
- [ ] ISE
- [X] VS Code
- [ ] Other (please specify)

#### Errors Received

```
PSMessageDetails      :
Exception             : System.Management.Automation.ExtendedTypeSystemException: The following exception occurred while trying to enumerate the collection: "Operation not supported on version 12.0.2000 SqlAzureDatabase.".
                         ---> Microsoft.SqlServer.Management.Sdk.Sfc.InvalidVersionEnumeratorException: Operation not supported on version 12.0.2000 SqlAzureDatabase.
                           at Microsoft.SqlServer.Management.Smo.XmlReadDoc.LoadInternal()
                           at Microsoft.SqlServer.Management.Smo.SqlObject.LoadInitDataFromAssemblyInternal(Assembly assemblyObject, String file, ServerVersion ver, String alias, StringCollection requestedFields, Boolean
                        store, StringCollection roAfterCreation, DatabaseEngineType databaseEngineType, DatabaseEngineEdition databaseEngineEdition, Stream configXml)
                           at Microsoft.SqlServer.Management.Smo.SqlObject.LoadInitDataFromAssembly(Assembly assemblyObject, String file, ServerVersion ver, DatabaseEngineType databaseEngineType, DatabaseEngineEdition
                        databaseEngineEdition)
                           at Microsoft.SqlServer.Management.Smo.SqlObject.LoadInitData(String file, ServerVersion ver, DatabaseEngineType databaseEngineType, DatabaseEngineEdition databaseEngineEdition)
                           at Microsoft.SqlServer.Management.Sdk.Sfc.ObjectCache.LoadElement(ObjectLoadInfo oli, ServerVersion ver, DatabaseEngineType databaseEngineType, DatabaseEngineEdition databaseEngineEdition)
                           at Microsoft.SqlServer.Management.Sdk.Sfc.ObjectCache.GetElement(ObjectLoadInfo oli, ServerVersion ver, DatabaseEngineType databaseEngineType, DatabaseEngineEdition databaseEngineEdition)
                           at Microsoft.SqlServer.Management.Sdk.Sfc.ObjectCache.GetAllElements(Urn urn, ServerVersion ver, DatabaseEngineType databaseEngineType, DatabaseEngineEdition databaseEngineEdition, Object ci)
                           at Microsoft.SqlServer.Management.Sdk.Sfc.Environment.GetObjectsFromCache(Urn urn, Object ci)
                           at Microsoft.SqlServer.Management.Sdk.Sfc.Environment.GetData(Request req, Object ci)
                           at Microsoft.SqlServer.Management.Sdk.Sfc.Enumerator.GetData(Object connectionInfo, Request request)
                           at Microsoft.SqlServer.Management.Smo.ExecutionManager.GetEnumeratorDataReader(Request req)
                           at Microsoft.SqlServer.Management.Smo.SqlSmoObject.InitChildLevel(Urn levelFilter, ScriptingPreferences sp, Boolean forScripting, IEnumerable`1 extraFields)
                           at Microsoft.SqlServer.Management.Smo.SmoCollectionBase.InitializeChildCollection(Boolean refresh, ScriptingPreferences sp, String filterQuery, IEnumerable`1 extraFields)
                           at Microsoft.SqlServer.Management.Smo.SmoCollectionBase.InitializeChildCollection(Boolean refresh, ScriptingPreferences sp)
                           at Microsoft.SqlServer.Management.Smo.SmoCollectionBase.GetEnumerator(ScriptingPreferences sp)
                           at Microsoft.SqlServer.Management.Smo.SmoCollectionBase.GetEnumerator()
                           at System.Management.Automation.EnumerableOps.GetEnumerator(IEnumerable enumerable)
                           --- End of inner exception stack trace ---
                           at System.Management.Automation.EnumerableOps.GetEnumerator(IEnumerable enumerable)
                           at System.Management.Automation.Interpreter.DynamicInstruction`2.Run(InterpretedFrame frame)
                           at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
                           at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
TargetObject          :
CategoryInfo          : NotSpecified: (:) [], ExtendedTypeSystemException
FullyQualifiedErrorId : ExceptionInGetEnumerator
ErrorDetails          :
InvocationInfo        : System.Management.Automation.InvocationInfo
ScriptStackTrace      : at Get-DbaServerRoleMember<Process>, C:\Users\burcha\Documents\PowerShell\Modules\dbatools\1.0.133\allcommands.ps1: line 34732
                        at Set-DbaLogin<Process>, C:\Users\burcha\Documents\PowerShell\Modules\dbatools\1.0.133\allcommands.ps1: line 67312
                        at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo : {}

```

#### Steps to Reproduce

```
$login = Get-DbaLogin -SqlInstance $SqlInstance -Login "LoginName"
$login | Set-DbaLogin -SecurePassword $SecurePassword
```

#### Expected Behavior

Password changed to new value - no errors raised

#### Actual Behavior

Password changed to new value, error raised as above

### Environmental information

```

PowerShell Version        : 7.0.3
dbatools latest installed : 1.0.133
Culture of OS             : en-GB

```

#### SQL Server:

```sql
Microsoft SQL Azure (RTM) - 12.0.2000.8
    Oct  1 2020 18:48:35
    Copyright (C) 2019 Microsoft Corporation
```

```sql
us_english
```

### PR DESCRIPTION
**Pull request by lancasteradam**
Get-DbaLogin, Set-DbaLogin, Connect-SqlInstance - add the -AzureUnsupported and more integration tests

- Added the -AzureUnsupported to Set-DbaLogin and Get-DbaLogin
- Updated Connect-SqlInstance to check the $SqlInstance.InputObject
- ~~Added param validation for Unlock in Set-DbaLogin~~ (moved to #7088)
- ~~Added logic to ensure the MustChange takes effect in Set-DbaLogin~~ (moved to #7089)
- ~~Added a new param to Set-DbaLogin for the check_expiration~~ (moved to #7090)
- Added integration tests

## Type of Change

 - [x] Bug fix (non-breaking change, fixes #7004 )
 - [x] New feature (non-breaking change, adds functionality, fixes # )
 - [ ] Breaking change (effects multiple commands or functionality, fixes # )
 - [x] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [x] Adding code coverage to existing functionality
 - [x] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/sqlcollaborative/appveyor-lab ?
 - [ ] Nunit test is included
 - [x] Documentation
 - [ ] Build system
 

### Purpose
 
Added the -AzureUnsupported to the commands and backfilled the integration tests.  ~~, and fixed some minor issues found by the integration tests.~~

### Commands to test

See the integration tests

### Learning
https://stackoverflow.com/questions/60117943/powershell-script-to-report-account-lockout-policy-settings

### CODE REVIEW

File: internal/functions/Connect-SqlInstance.ps1 [hunk 1]
```diff
@@ -66,7 +66,7 @@ function Connect-SqlInstance {
         [switch]$NonPooled
     )
     if ($SqlInstance.InputObject.GetType().Name -eq 'Server') {
-        if ($AzureUnsupported -and $server.DatabaseEngineType -eq "SqlAzureDatabase") {
+        if ($AzureUnsupported -and $SqlInstance.InputObject.DatabaseEngineType -eq "SqlAzureDatabase") {
```
Comment (wsmelton) [inline internal/functions/Connect-SqlInstance.ps1:69]: Not sure I see the reason to check the input object? It is known since this is an internal command where we only support the `-AzureUnsupported` when we know it can't use it.

File: functions/Set-DbaLogin.ps1 [hunk 2]
```diff
@@ -253,8 +270,20 @@ function Set-DbaLogin {
                 # Change the password
                 if (Test-bound -ParameterName 'SecurePassword') {
                     try {
+                        # these settings are required for must_change and are validated above
+                        if (Test-Bound MustChange) {
+                            Write-Message -Message "Setting check_policy and check_expiration to true for $l because the must_change option requires it." -Level Warning
```
Comment (wsmelton) [inline functions/Set-DbaLogin.ps1:275]: This is not the purpose of providing MustChange in the command the way I read it. Enabling these settings breaks the use of that parameter's purpose. It is meant to be used in conjunction with a new password being provided so it will force a change in the password.

Out intention in that workflow is not to enforce any policies; having the password policy enforced can cause the password change itself to fail if they are using a weak password.

File: functions/Get-DbaLogin.ps1 [hunk 3]
```diff
@@ -20,7 +20,7 @@ function Get-DbaLogin {
         The login(s) to process - this list is auto-populated from the server. If unspecified, all logins will be processed.
 
     .PARAMETER ExcludeLogin
-        The login(s) to exclude - this list is auto-populated from the server
+        The login(s) to exclude
```
Comment (wsmelton) [inline functions/Get-DbaLogin.ps1:23]: Please revert this change.

**Review by wsmelton — CHANGES_REQUESTED** [review]
Changes for the specified issue can be put in the PR. The other changes need an issue created for further discussion before making the changes.

File: functions/Set-DbaLogin.ps1 [hunk 4]
```diff
@@ -209,6 +213,19 @@ function Set-DbaLogin {
...
+            Stop-Function -Message 'You must specify a password when using the -Unlock parameter'
+        }
+
+        if ((Test-Bound MustChange) -and (Test-Bound SecurePassword -Not)) {
+            Stop-Function -Message 'You must specify a password when using the -MustChange parameter'
+        }
+
+        # check_expiration and check_policy are required for must_change
+        if ((Test-Bound MustChange) -and ((Test-Bound PasswordPolicyEnforced -Not) -or (Test-Bound PasswordExpirationEnabled -Not))) {
+            Stop-Function -Message 'The -MustChange parameter must be used with -PasswordPolicyEnforced and -PasswordExpirationEnabled'
```
Comment (wsmelton) [inline functions/Set-DbaLogin.ps1:227]: This will break intended functionality of the command.

File: functions/Set-DbaLogin.ps1 [hunk 5]
```diff
@@ -253,8 +270,20 @@ function Set-DbaLogin {
                 # Change the password
                 if (Test-bound -ParameterName 'SecurePassword') {
                     try {
+                        # these settings are required for must_change and are validated above
+                        if (Test-Bound MustChange) {
+                            Write-Message -Message "Setting check_policy and check_expiration to true for $l because the must_change option requires it." -Level Warning
+                            $l.PasswordPolicyEnforced = $PasswordPolicyEnforced
+                            $l.PasswordExpirationEnabled = $PasswordExpirationEnabled
+                            $l.Alter()
```
Comment (wsmelton) [inline functions/Set-DbaLogin.ps1:278]: This can cause unintended changes that may not be desired.

Comment (wsmelton) [reply functions/Set-DbaLogin.ps1:278]: The MustChange is meant for a specific use-case, we are not accounting for the other use case in the command. An issue needs to be opened for any bug surrounding the parameter as-is to better discuss how to address it.

**Comment by lancasteradam** [comment]
ok, sounds good. I'll simplify this PR and open a new issue for the other items.

**Comment by lancasteradam** [comment]
These issues were created for the other proposed changes:

https://github.com/sqlcollaborative/dbatools/issues/7088
https://github.com/sqlcollaborative/dbatools/issues/7089
https://github.com/sqlcollaborative/dbatools/issues/7090

File: functions/Get-DbaLogin.ps1 [hunk 6]
```diff
@@ -20,7 +20,7 @@ function Get-DbaLogin {
         The login(s) to process - this list is auto-populated from the server. If unspecified, all logins will be processed.
 
     .PARAMETER ExcludeLogin
-        The login(s) to exclude - this list is auto-populated from the server
+        The login(s) to exclude from an auto-populated list of logins from the server.
```
Comment (wsmelton) [inline functions/Get-DbaLogin.ps1:23]:
```suggestion
        The login(s) to exclude. Options for this list are auto-populated from the server.
```

This matches the text we use in other commands for the same parameter (sync and copy commands).

File: functions/Get-DbaLogin.ps1 [hunk 7]
```diff
@@ -1,7 +1,7 @@
 function Get-DbaLogin {
     <#
     .SYNOPSIS
-        Function to get an SMO login object of the logins for a given SQL Server instance. Takes a server object from the pipeline.
+        Function to get an SMO login object of the logins for a given SQL Server instance. Takes a server object from the pipeline. SQL Azure DB is not supported.
```
Comment (wsmelton) [inline functions/Get-DbaLogin.ps1:4]:
```suggestion
        Function to get an SMO login object of the logins for a given SQL Server instance. Takes a server object from the pipeline.
        SQL Azure DB is not supported.
```

Adding a line break will give this more attention

File: functions/Set-DbaLogin.ps1 [hunk 8]
```diff
@@ -1,7 +1,7 @@
 function Set-DbaLogin {
     <#
     .SYNOPSIS
-        Set-DbaLogin makes it possible to make changes to one or more logins.
+        Set-DbaLogin makes it possible to make changes to one or more logins. SQL Azure DB is not supported.
```
Comment (wsmelton) [inline functions/Set-DbaLogin.ps1:4]:
```suggestion
        Set-DbaLogin makes it possible to make changes to one or more logins.
        SQL Azure DB is not supported.
```

**Review by wsmelton — CHANGES_REQUESTED** [review]

**Comment by wsmelton** [comment]
Please update the title (if needed) and body of PR

**Review by wsmelton — APPROVED** [review]
